sklearn/utils/_array_api.py

```diff
@@ -221,49 +322,42 @@ def get_namespace(*arrays):
     Returns
     -------
     namespace : module
-        Namespace shared by array objects.
+        Namespace shared by array objects. If any of the `arrays` are not arrays,
+        the namespace defaults to NumPy.
 
-    is_array_api : bool
+    is_array_api_compliant : bool
         True of the arrays are containers that implement the Array API spec.
+        Always False when array_api_dispatch=False.
     """
-    # `arrays` contains one or more arrays, or possibly Python scalars (accepting
-    # those is a matter of taste, but doesn't seem unreasonable).
-    # Returns a tuple: (array_namespace, is_array_api)
-
-    if not get_config()["array_api_dispatch"]:
-        return _NumPyAPIWrapper(), False
+    array_api_dispatch = get_config()["array_api_dispatch"]
+    if not array_api_dispatch:
+        return _NUMPY_API_WRAPPER_INSTANCE, False
 
-    namespaces = {
-        x.__array_namespace__() if hasattr(x, "__array_namespace__") else None
-        for x in arrays
-        if not isinstance(x, (bool, int, float, complex))
-    }
+    _check_array_api_dispatch(array_api_dispatch)
 
-    if not namespaces:
-        # one could special-case np.ndarray above or use np.asarray here if
-        # older numpy versions need to be supported.
-        raise ValueError("Unrecognized array input")
+    # array-api-compat is a required dependency of scikit-learn only when
+    # configuring `array_api_dispatch=True`. Its import should therefore be
+    # protected by _check_array_api_dispatch to display an informative error
+    # message in case it is missing.
+    import array_api_compat
 
-    if len(namespaces) != 1:
-        raise ValueError(f"Multiple namespaces for array inputs: {namespaces}")
+    namespace, is_array_api_compliant = array_api_compat.get_namespace(*arrays), True
 
-    (xp,) = namespaces
-    if xp is None:
-        # Use numpy as default
-        return _NumPyAPIWrapper(), False
+    if namespace.__name__ in {"numpy.array_api", "cupy.array_api"}:
```

Comment:
Shouldn't we wrap all(?) array API namespaces in the `_ArrayAPIWrapper`? This means I'd add `cupy` and `torch` here, but not `numpy`.

It took me a while to get this straight in my head, so maybe it is still confused but here is how I understand it:

`_NumPyAPIWrapper` is meant to make "normal" and older numpy arrays look like array API arrays. It adds  things to the Numpy namespace that are present in an Array API namespace but are missing (or different) in "normal" Numpy.

`_ArrayAPIWrapper` is meant to wrap fully Array API compatible namespaces and smooth out differences between different Array API implementations (say, cupy vs torch) as well as adding things that aren't yet in the Array API or have changed between Array API versions.

If yes, shouldn't we wrap all "not Numpy" namespaces in `_ArrayAPIWrapper` and maybe in the future have different wrappers for different namespaces (say there is a bug in the torch array API implementation that we want to patch).

And taking this one step further: in https://github.com/scikit-learn/scikit-learn/issues/26024#issuecomment-1515994477 I am looking for a way to deal with the fact that Numpy offers `nanmin()` but the array API does not. I think a natural way of dealing with this would be to add an implementation of `nanmin` to `_ArrayAPIWrapper` so that the scikit-learn code can keep using `xp.nanmin()`.

WDYT?